javalib/src/main/scala/java/net/AbstractPlainSocketImpl.scala

```diff
@@ -454,25 +479,34 @@ private[net] abstract class AbstractPlainSocketImpl extends SocketImpl {
 
         ptr.asInstanceOf[Ptr[Byte]]
       case SocketOptions.SO_TIMEOUT =>
-        val ptr = stackalloc[timeval]
         val mseconds = value.asInstanceOf[Int]
-
         this.timeout = mseconds
 
-        ptr.tv_sec = mseconds / 1000
-        ptr.tv_usec = (mseconds % 1000) * 1000
+        if (isWindows) {
+          val ptr = stackalloc[DWord]
+          !ptr = mseconds.toUInt
+          ptr.asInstanceOf[Ptr[Byte]]
+        } else {
+          val ptr = stackalloc[timeval]
 
-        ptr.asInstanceOf[Ptr[Byte]]
+          ptr.tv_sec = mseconds / 1000
+          ptr.tv_usec = (mseconds % 1000) * 1000
+
+          ptr.asInstanceOf[Ptr[Byte]]
+        }
       case _ =>
         val ptr = stackalloc[CInt]
         !ptr = value.asInstanceOf[Int]
         ptr.asInstanceOf[Ptr[Byte]]
     }
 
-    if (socket.setsockopt(fd.fd, level, optValue, opt, len) == -1) {
+    if (socket.setsockopt(fd.fd, level, optValue, opt, len) != 0) {
+      val errCode =
+        if (isWindows) WSAGetLastError()
+        else errno.errno
```

Comment:
This fragment is repeated a few times. Consider putting it in a reusable private def.